sdk/core/Azure.Core/src/ResourceIdentifier.cs

```diff
@@ -0,0 +1,429 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using System.Text;
+
+namespace Azure.Core
+{
+    /// <summary>
+    /// An Azure Resource Manager resource identifier.
+    /// </summary>
+    public sealed class ResourceIdentifier : IEquatable<ResourceIdentifier>, IComparable<ResourceIdentifier>
+    {
+        private const char Separator = '/';
+
+        private static ResourceType TenantResourceType = new ResourceType("Microsoft.Resources/tenants");
+        private static ResourceType SubscriptionResourceType = new ResourceType("Microsoft.Resources/subscriptions");
+        private static ResourceType ResourceGroupResourceType = new ResourceType("Microsoft.Resources/resourceGroups");
+        private static ResourceType ProviderResourceType = new ResourceType("Microsoft.Resources/providers");
+        private const string RootStringValue = "/";
+        private const string ProvidersKey = "providers";
+        private const string SubscriptionsKey = "subscriptions";
+        private const string LocationsKey = "locations";
+        private const string ResourceGroupsLowerKey = "resourcegroups";
+
+        private string? _stringValue;
+
+        /// <summary>
+        /// The root of the resource hierarchy.
+        /// </summary>
+        public static readonly ResourceIdentifier Root = new ResourceIdentifier(null, TenantResourceType, string.Empty, true);
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ResourceIdentifier"/> class.
+        /// </summary>
+        /// <param name="resourceId"> The id string to create the ResourceIdentifier from. </param>
+        public ResourceIdentifier(string resourceId)
+        {
+            var id = Create(resourceId);
+            Init(id.Parent, id.ResourceType, id.Name);
+            ResourceType = id.ResourceType;
```

Comment:
Curious why these assignments don't also happen in init, since they happen in both constructors?

Reply:
because they are no nullable and so  the compiler requires it CS 8618

Reply:
Per offline discussion - we had discussed disabling the compiler warning, but I think that might be unhelpful to future editors of this class.  If it's possible to add a comment saying this is duplicated across constructors due to this compiler warning, I think that'd be ideal.